File: android/src/main/kotlin/com/klarna/inapp/sdk/flutter_klarna_inapp_sdk/postpurchaseexperience/PostPurchaseExperienceHandler.kt
package com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.postpurchaseexperience

import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.handler.BaseMethodHandler
import io.flutter.plugin.common.MethodChannel

internal object PostPurchaseExperienceHandler : BaseMethodHandler<PostPurchaseExperienceMethod>(PostPurchaseExperienceMethod.Parser) {

    val ppeManagerMap = mutableMapOf<Int, PostPurchaseExperienceManager>()

    override fun onMethod(method: PostPurchaseExperienceMethod, result: MethodChannel.Result) {
        when (method) {
            is PostPurchaseExperienceMethod.Initialize -> initialize(method, result)
            is PostPurchaseExperienceMethod.Destroy -> destroy(method, result)
            is PostPurchaseExperienceMethod.RenderOperation -> renderOperation(method, result)
            is PostPurchaseExperienceMethod.AuthorizationRequest -> authorizationRequest(method, result)
        }
    }

    private fun initialize(method: PostPurchaseExperienceMethod.Initialize, result: MethodChannel.Result) {
        val manager = getManager(method)
        if (manager == null) {
            val newManager = PostPurchaseExperienceManager()
            newManager.initialize(method, result)
            ppeManagerMap[method.id] = newManager
        } else {
            getManager(method)?.initialize(method, result)
        }
    }

    private fun destroy(method: PostPurchaseExperienceMethod.Destroy, result: MethodChannel.Result) {
        getManager(method)?.destroy(method, result)
        ppeManagerMap.remove(method.id)
    }

    private fun renderOperation(method: PostPurchaseExperienceMethod.RenderOperation, result: MethodChannel.Result) {
        getManager(method)?.renderOperation(method, result)
    }

    private fun authorizationRequest(method: PostPurchaseExperienceMethod.AuthorizationRequest, result: MethodChannel.Result) {
        getManager(method)?.authorizationRequest(method, result)
    }

    private fun getManager(method: PostPurchaseExperienceMethod): PostPurchaseExperienceManager? {
        return ppeManagerMap[method.id]
    }
}

File: ios/Classes/PostPurchaseExperience/PostPurchaseExperienceHandler.swift
import Flutter
import WebKit
import KlarnaMobileSDK

class PostPurchaseExperienceHandler: BaseMethodHandler<PostPurchaseExperienceMethod> {
    
    private var ppeManagerMap = [Int:PostPurchaseExperienceManager]()
    
    init() {
        super.init(parser: PostPurchaseExperienceMethods.Parser())
    }
    
    override func onMethod(method: PostPurchaseExperienceMethod, result: @escaping FlutterResult) {
        switch method {
        case is PostPurchaseExperienceMethods.Initialize:
            initialize(method: method as! PostPurchaseExperienceMethods.Initialize, result: result)
        case is PostPurchaseExperienceMethods.Destroy:
            destroy(method: method as! PostPurchaseExperienceMethods.Destroy, result: result)
        case is PostPurchaseExperienceMethods.RenderOperation:
            renderOperation(method: method as! PostPurchaseExperienceMethods.RenderOperation, result: result)
        case is PostPurchaseExperienceMethods.AuthorizationRequest:
            authorizationRequest(method: method as! PostPurchaseExperienceMethods.AuthorizationRequest, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initialize(method: PostPurchaseExperienceMethods.Initialize, result: @escaping FlutterResult) {
        if let manager = getManager(method: method) {
            manager.initialize(method: method, result: result)
        } else {
            let manager = PostPurchaseExperienceManager()
            manager.initialize(method: method, result: result)
            ppeManagerMap[method.id] = manager
        }
    }
    
    private func destroy(method: PostPurchaseExperienceMethods.Destroy, result: @escaping FlutterResult) {
        getManager(method: method)?.destroy(method: method, result: result)
        ppeManagerMap.removeValue(forKey: method.id)
    }
    
    private func renderOperation(method: PostPurchaseExperienceMethods.RenderOperation, result: @escaping FlutterResult) {
        getManager(method: method)?.renderOperation(method: method, result: result)
    }
    
    private func authorizationRequest(method: PostPurchaseExperienceMethods.AuthorizationRequest, result: @escaping FlutterResult) {
        getManager(method: method)?.authorizationRequest(method: method, result: result)
    }
    
    private func getManager(method: PostPurchaseExperienceMethod) -> PostPurchaseExperienceManager? {
        return ppeManagerMap[method.id]
    }
    
}
